daphne_server/src/roles/aggregator.rs

```diff
@@ -118,18 +287,93 @@ impl<S: Sync> DapAggregator<S> for crate::App {
+            .await?
+            .ok_or(DapError::Abort(DapAbort::UnrecognizedTask))?;
+
+        // Check whether the request overlaps with previous requests. This is done by
+        // checking the AggregateStore and seeing whether it requests for aggregate
+        // shares that have already been marked collected.
+        let durable = self.durable();
+        let mut requests = Vec::new();
+        for bucket in task_config.as_ref().batch_span_for_sel(batch_sel)? {
+            requests.push(
+                durable
+                    .request(
+                        bindings::AggregateStore::CheckCollected,
+                        (task_config.as_ref().version, &task_id.to_hex(), &bucket),
+                    )
+                    .send(),
+            );
+        }
+
+        let responses: Vec<bool> = try_join_all(requests)
             .await
+            .map_err(|e| fatal_error!(err = ?e))?;
+
+        for collected in responses {
+            if collected {
+                return Ok(true);
+            }
+        }
+
+        Ok(false)
     }
 
     async fn batch_exists(&self, task_id: &TaskId, batch_id: &BatchId) -> Result<bool, DapError> {
-        todo!()
+        let task_config = self
+            .get_task_config_for(task_id)
+            .await?
+            .ok_or(DapError::Abort(DapAbort::UnrecognizedTask))?;
+
+        let agg_share: DapAggregateShare = self
+            .durable()
+            .request(
+                bindings::AggregateStore::Get,
+                (
+                    task_config.as_ref().version,
+                    &task_id.to_hex(),
+                    &DapBatchBucket::FixedSize {
+                        batch_id: *batch_id,
+                    },
+                ),
+            )
+            .send()
+            .await
+            .map_err(|e| fatal_error!(err = ?e))?;
+
+        Ok(!agg_share.empty())
     }
 
     async fn current_batch(&self, task_id: &TaskId) -> Result<BatchId, DapError> {
-        todo!()
+        let task_config = self
+            .get_task_config_for(task_id)
+            .await?
+            .ok_or(DapError::Abort(DapAbort::UnrecognizedTask))?;
+
+        if !matches!(task_config.query, DapQueryConfig::FixedSize { .. }) {
+            return Err(fatal_error!(err = "query type mismatch"));
+        }
+
+        let res: LeaderBatchQueueResult = self
+            .durable()
+            .request(
+                bindings::LeaderBatchQueue::Current,
+                (task_config.as_ref().version, &task_id.to_hex()),
+            )
+            .send()
+            .await
+            .map_err(|e| fatal_error!(err = ?e))?;
+
+        match res {
+            LeaderBatchQueueResult::Ok(batch_id) => Ok(batch_id),
+
+            // TODO spec: If we end up taking the current batch semantics of
+            // https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/313, then we'll need to
+            // define an error type for this case.
```

Comment:
This has been settled.
```suggestion
            // TODO Return 400 Bad Request (with problem detail "empty batch queue") here instead 500 Internal Error
```